example/offload-template/src/StepDiagnostic.cc

```diff
@@ -0,0 +1,161 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/user/StepDiagnostic.cc
+//---------------------------------------------------------------------------//
+#include "StepDiagnostic.hh"
+
+#include <type_traits>
+#include <vector>
+#include <celeritas/global/ActionInterface.hh>
+#include <celeritas/global/ActionLauncher.hh>
+#include <celeritas/global/CoreParams.hh>
+#include <celeritas/global/CoreState.hh>
+#include <celeritas/global/TrackExecutor.hh>
+#include <celeritas/track/CoreStateCounters.hh>
+#include <corecel/Macros.hh>
+#include <corecel/data/AuxParamsRegistry.hh>
+#include <corecel/data/AuxStateVec.hh>
+#include <corecel/data/Collection.hh>
+#include <corecel/data/CollectionAlgorithms.hh>
+#include <corecel/data/PinnedAllocator.hh>
+#include <corecel/io/Logger.hh>
+#include <corecel/sys/ActionRegistry.hh>
+#include <geocel/g4/Convert.hh>
+
+#include "StepDiagnosticExecutor.hh"
+
+namespace celeritas
+{
+namespace example
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct and add to core params.
+ */
+std::shared_ptr<StepDiagnostic>
+StepDiagnostic::make_and_insert(CoreParams const& core)
+{
+    ActionRegistry& actions = *core.action_reg();
+    AuxParamsRegistry& aux = *core.aux_reg();
+    auto result
+        = std::make_shared<StepDiagnostic>(actions.next_id(), aux.next_id());
+    actions.insert(result);
+    aux.insert(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with IDs and filename base.
+ *
+ * This also writes to the "metadata" JSON suffix.
+ */
+StepDiagnostic::StepDiagnostic(ActionId action_id, AuxId aux_id)
+    : sad_{action_id, "step-diagnostic", "accumulate step statistics"}
+    , aux_id_{aux_id}
+{
+    CELER_EXPECT(aux_id_);
+
+    // Set up shared data on host and device
+    mirror_ = CollectionMirror{HostVal<StepParamsData>{}};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the statistics and reset them.
+ */
+StepStatistics StepDiagnostic::GetAndReset(CoreStateInterface& state) const
+{
+    // Kernel-collected statistics copied to host memory
+    NativeStepStatistics data;
+    HostStepStatistics host_data;
+
+    auto try_copy = [&](auto* core_state) -> bool {
```

Comment:
Is it just me or this seems like it could be a separate function? Since the `StepDiagnostic` will be abstracted in the future, seems like this (fetch step state data for the user) is already a useful abstraction "placeholder".